### PR TITLE
Implement Multi-GPU support for Stable Diffusion finetuner

### PR DESCRIPTION
**Pull request by harubaru**

### CODE REVIEW

File: sd-finetuner-workflow/sd-finetune-workflow.yaml [hunk 1]
```diff
@@ -106,9 +108,9 @@ spec:
     - name: downloader_tag
       value: '797b903'
     - name: finetuner_image
-      value: 'ghcr.io/coreweave/ml-containers/sd-finetuner'
+      value: 'harubaru1/sd-finetuner'
     - name: finetuner_tag
-      value: '1924bc5'
+      value: 'r9'
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetune-workflow.yaml:113]: Build the Docker image as CI, then update this.

File: sd-finetuner-workflow/sd-finetune-workflow.yaml [hunk 2]
```diff
@@ -268,7 +273,10 @@ spec:
         - name: project_id
     container:
       image: "{{workflow.parameters.finetuner_image}}:{{workflow.parameters.finetuner_tag}}"
-      command: [ "/usr/bin/python3", "/app/finetuner.py" ]
+      command: [ "accelerate-launch",
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetune-workflow.yaml:276]: Which python is this `accelerate-launch` using?

File: sd-finetuner-workflow/sd-finetuner/Dockerfile [hunk 3]
```diff
@@ -7,7 +7,8 @@ RUN apt-get install -y cuda-nvcc-11-3 cuda-nvml-dev-11-3 libcurand-dev-11-3 \
 RUN mkdir /app
 WORKDIR /app
 COPY requirements.txt .
+COPY accelerate_config.yaml .
 RUN pip3 install --no-cache-dir -r requirements.txt
 RUN pip3 install --no-cache-dir torch torchvision --extra-index-url https://download.pytorch.org/whl/cu116 --upgrade
 COPY finetuner.py .
-CMD [ "/usr/bin/python3", "finetuner.py" ]
+CMD [ "accelerate-launch", "finetuner.py" ]
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetuner/Dockerfile:14]: Is this correctly in the path? Docker images usually should reference the entire path for the executable.

File: sd-finetuner-workflow/sd-finetuner/finetuner.py [hunk 4]
```diff
@@ -452,34 +440,34 @@ def __init__(
...
-            project=args.project_id,
-            name=args.run_name,
-            config={
-                k: v for k, v in vars(args).items() if k not in ["hf_token"]
-            },
-            dir=args.output_path + "/wandb",
-        )
-        self.global_step = 0
+        if accelerator.is_main_process:
+            self.progress_bar = tqdm.tqdm(
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetuner/finetuner.py:448]: `tqdm` is pretty terrible for Argo workflows and logging. Look at the finetuner for how I work with it.

File: sd-finetuner-workflow/sd-finetuner/finetuner.py [hunk 5]
```diff
@@ -526,62 +514,75 @@ def sample(self, prompt: str) -> None:
...
-            device=latents.device,
-        )
-        timesteps = timesteps.long()
+        train_loss = 0.0
+        with self.accelerator.accumulate(self.unet):
+            # Convert images to latent space
+            latents = self.vae.encode(
+                batch["pixel_values"].to(self.weight_dtype)
+            ).latent_dist.sample()
+            latents = latents * 0.18215
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetuner/finetuner.py:523]: What is this magic number? Should this be a constant?

File: sd-finetuner-workflow/sd-finetuner/finetuner.py [hunk 6]
```diff
@@ -526,62 +514,75 @@ def sample(self, prompt: str) -> None:
...
+
+            # Sample noise
+            noise = torch.randn_like(latents)
```
Comment (wbrown) [inline sd-finetuner-workflow/sd-finetuner/finetuner.py:526]: Do we know if this uses the seed set by accelerate?

Comment (harubaru) [reply sd-finetuner-workflow/sd-finetuner/finetuner.py:526]: Yep, accelerate sets the Torch seed with it https://github.com/huggingface/accelerate/blob/21b7f15c960ff61fe0ff976d74c715622f5c2b00/src/accelerate/utils/random.py#L44

File: sd-finetuner-workflow/sd-finetuner/finetuner.py (hunk 5 above)
Comment (harubaru) [reply sd-finetuner-workflow/sd-finetuner/finetuner.py:523]: It is a constant used for scaling the latents https://github.com/huggingface/diffusers/issues/437, the original purpose being to handle different latent spaces from different autoencoders.

**Comment by harubaru** [comment]
This PR would need docs updated to inform the user on how to enable multi-gpu training which would be a minor addition, as it would only involve updating the `trainer_gpu_count` parameter in the workflow

**Review by wbrown — COMMENTED** [review]
:+1:
